write a guide for routes in multiple namespaces (#452)
kubernetes-sigs/service-apis: docs-src/multiple-ns.md

# Routes in multiple namespaces

With the Service APIs, a single Gateway can target routes across multiple
namespaces. 

This guide assumes that you have installed Service APIs CRDs and a conformant
controller.

In the following example:

- `acme-lb` GatewayClass: The GatewayClass responsible for satisfying Gateway
  and Route resources.
- `multi-ns-gateaway` Gateway: The Gateway is configured with a single listener

>   on port 80 which selects routes from **All namespaces** which have the label

Reviewer: Nit: "All" -> "all"

  `product: baz`. Notice how the `routes.namespaces.from` field in the listener
  is set to `All`.

> - `foo-ns` and `bar-ns` Namespaces: These are two namespaces in which we route

Reviewer: I'm not exactly sure where this belongs, but it would probably be good to call attention to the fact that this example creates namespaces and explicitly installs resources in them. If a user is trying to keep all their examples in a single namespace, that will not work here. Similarly, if they happen to be using namespaces with these names that could be problematic. Especially if they happen to cleanup demo resources with `kubectl delete -f`. 

Maybe to be extra safe resources like namespaces should have a clear association with this project? Something like `service-apis-example-ns1`, `service-apis-example-ns2` is overly verbose but also unlikely to conflict with any existing namespaces. Maybe there's a less verbose option that has a similar effect of clearly being tied to an example from this project?

Author: Opted to change the name to not conflict with anything that already exists.
I'm not explicitly calling it out because it seems self-explanatory. Is that okay?

  resources are installed.
- `http-app-1` and `http-app-2` HTTPRoutes: These are two resources that are
  installed in separate namespaces. These routes will be picked up by
  `multi-ns-gateway` resource for two reasons:
    - Both have the `product: baz` label on them.
    - Both have `spec.gateways.allow` set to `All`. The route owner needs to
      explictly set this field so that Gateways from other namespaces can target

>       the route.

Reviewer: Can we use this example to show both approaches? One with `allow: All` and one with `allow: FromList`. Unfortunately that would also require creating a namespace for the Gateway.

Author: > Unfortunately that would also require creating a namespace for the Gateway.

I'll put the Gateway in the default namespace.

# Routes in multiple namespaces

With the Service APIs, a single Gateway can target routes across multiple
namespaces. 

This guide assumes that you have installed Service APIs CRDs and a conformant
controller.

In the following example:

- `acme-lb` GatewayClass: The GatewayClass responsible for satisfying Gateway
  and Route resources.
- `multi-ns-gateaway` Gateway: The Gateway is configured with a single listener

>   on port 80 which selects routes from **all namespaces** which have the label
>   `product: baz`. Notice how the `routes.namespaces.from` field in the listener

Reviewer: This is a bit confusing - it almost sounds like the namespaces need to have those labels, not the routes. 

```suggestion
  on port 80 which selects routes that have the label `product: baz` in any namespace. 
  Notice how the `routes.namespaces.from` field in the listener
```

  is set to `All`.
- `service-apis-example-ns1` and `service-apis-example-ns2` Namespaces: These

>   are two namespaces in which we route resources are installed.

Reviewer 2: s/These are two namespaces in which we route resources are installed/These are the namespaces in which route resources are instantiated.

- `http-app-1` and `http-app-2` HTTPRoutes: These are two resources that are
  installed in separate namespaces. These routes will be picked up by

>   `multi-ns-gateway` resource for the following reasons:

Reviewer 2: s/These routes will be picked up by `multi-ns-gateway` resource for the following reasons/These routes will be bound to Gateway `multi-ns-gateway` for the following reasons/

    - Both have the `product: baz` label on them.
    - `http-app-1` HTTPRoute has `spec.gateways.allow` set to `All`.  The route
      owner here has opted to allow **all** Gateways in the cluster to bind to

>       this Route.

Reviewer 2: The route owner ~~here~~ has opted to allow **all** Gateways in the cluster to bind to this Route.

    - `http-app-2` HTTPRoute has `spec.gateways.allow` set to `FromList` and
      contains a reference to the `multi-ns-gateway` in `default` namespace. In

>       this case, no other Gateway can bind to this route.  explictly set this

Reviewer: ```suggestion
      this case, only Gateways in the `default` namespace can bind to this route.
```

>       field so that Gateways from other namespaces can target the route.

Reviewer: Maybe something like "Additional namespaces may be added to this list to allow Gateways in other namespaces to reference this namespace."

Author: This was worded completely wrong. Updated to rephrase and include your suggestion.
